File: backend/app/services/email_security_service.py
import logging
import hashlib
import requests
from datetime import datetime, timedelta
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.models import MonitoredEmail, EmailSecurityCheck, EmailSecurityAlert
from app.schemas import MonitoredEmailCreate, MonitoredEmailUpdate

logger = logging.getLogger(__name__)


class EmailSecurityService:
    def __init__(self, db: Session):
        self.db = db
        self.haveibeenpwned_api = "https://api.pwnedpasswords.com/range/"

    # --- Monitored Emails ---
    def create_monitored_email(self, email: MonitoredEmailCreate, user_id: int) -> MonitoredEmail:
        """Ajouter un email à surveiller"""
        db_email = MonitoredEmail(**email.dict(), user_id=user_id)
        self.db.add(db_email)
        self.db.commit()
        self.db.refresh(db_email)
        
        # Perform initial check
        self.check_email_security(db_email.id, user_id)
        
        return db_email

    def get_monitored_emails(self, user_id: int, active_only: bool = False) -> List[MonitoredEmail]:
        """Liste des emails surveillés"""
        query = self.db.query(MonitoredEmail).filter(MonitoredEmail.user_id == user_id)
        
        if active_only:
            query = query.filter(MonitoredEmail.active == True)
        
        return query.order_by(MonitoredEmail.created_at.desc()).all()

    def get_monitored_email(self, email_id: int, user_id: int) -> Optional[MonitoredEmail]:
        """Obtenir un email surveillé"""
        return self.db.query(MonitoredEmail).filter(
            MonitoredEmail.id == email_id,
            MonitoredEmail.user_id == user_id
        ).first()

    def update_monitored_email(self, email_id: int, email: MonitoredEmailUpdate, user_id: int) -> Optional[MonitoredEmail]:
        """Mettre à jour un email surveillé"""
        db_email = self.get_monitored_email(email_id, user_id)
        if not db_email:
            return None
        
        for key, value in email.dict(exclude_unset=True).items():
            setattr(db_email, key, value)
        
        self.db.commit()
        self.db.refresh(db_email)
        return db_email

    def delete_monitored_email(self, email_id: int, user_id: int) -> bool:
        """Supprimer un email surveillé"""
        db_email = self.get_monitored_email(email_id, user_id)
        if not db_email:
            return False
        
        self.db.delete(db_email)
        self.db.commit()
        return True

    # --- Security Checks ---
    def check_email_security(self, email_id: int, user_id: int) -> Optional[EmailSecurityCheck]:
        """Vérifier la sécurité d'un email via HaveIBeenPwned"""
        db_email = self.get_monitored_email(email_id, user_id)
        if not db_email:
            return None
        
        try:
            # Use k-anonymity for privacy
            email_hash = hashlib.sha1(db_email.email.lower().encode()).hexdigest().upper()
            prefix = email_hash[:5]
            suffix = email_hash[5:]
            
            # Check HaveIBeenPwned API
            response = requests.get(f"{self.haveibeenpwned_api}{prefix}", timeout=10)
            
            if response.status_code == 200:
                breaches = []
                breach_count = 0
                
                for line in response.text.splitlines():
                    if line.startswith(suffix):
                        count = int(line.split(':')[1])
                        breach_count += count
                        breaches.append({
                            "hash_suffix": suffix,
                            "count": count
                        })
                
                is_compromised = breach_count > 0
                
                # Create check record
                check = EmailSecurityCheck(
                    monitored_email_id=email_id,
                    is_compromised=is_compromised,
                    breach_count=breach_count,
                    breach_details=breaches if breaches else None
                )
                
                self.db.add(check)
                self.db.commit()
                self.db.refresh(check)
                
                # Create alert if compromised
                if is_compromised:
                    self.create_alert(check.id)
                
                logger.info(f"Security check completed for {db_email.email}: {breach_count} breaches")
                return check
                
            else:
                logger.error(f"HaveIBeenPwned API error: {response.status_code}")
                return None
                
        except Exception as e:
            logger.error(f"Error checking email security for {db_email.email}: {e}")
            return None

    def get_security_checks(self, user_id: int, email_id: Optional[int] = None, limit: int = 50) -> List[EmailSecurityCheck]:
        """Historique des vérifications"""
        query = self.db.query(EmailSecurityCheck).join(MonitoredEmail).filter(
            MonitoredEmail.user_id == user_id
        )
        
        if email_id:
            query = query.filter(EmailSecurityCheck.monitored_email_id == email_id)
        
        return query.order_by(EmailSecurityCheck.checked_at.desc()).limit(limit).all()

    def get_latest_check(self, email_id: int, user_id: int) -> Optional[EmailSecurityCheck]:
        """Dernière vérification pour un email"""
        return self.db.query(EmailSecurityCheck).join(MonitoredEmail).filter(
            EmailSecurityCheck.monitored_email_id == email_id,
            MonitoredEmail.user_id == user_id
        ).order_by(EmailSecurityCheck.checked_at.desc()).first()

    # --- Alerts ---
    def create_alert(self, check_id: int) -> EmailSecurityAlert:
        """Créer une alerte de sécurité"""
        alert = EmailSecurityAlert(check_id=check_id)
        self.db.add(alert)
        self.db.commit()
        self.db.refresh(alert)
        return alert

    def get_alerts(self, user_id: int, acknowledged: Optional[bool] = None) -> List[EmailSecurityAlert]:
        """Liste des alertes"""
        query = self.db.query(EmailSecurityAlert).join(EmailSecurityCheck).join(MonitoredEmail).filter(
            MonitoredEmail.user_id == user_id
        )
        
        if acknowledged is not None:
            query = query.filter(EmailSecurityAlert.acknowledged == acknowledged)
        
        return query.order_by(EmailSecurityAlert.sent_at.desc()).all()

    def acknowledge_alert(self, alert_id: int, user_id: int) -> Optional[EmailSecurityAlert]:
        """Acquitter une alerte"""
        alert = self.db.query(EmailSecurityAlert).join(EmailSecurityCheck).join(MonitoredEmail).filter(
            EmailSecurityAlert.id == alert_id,
            MonitoredEmail.user_id == user_id
        ).first()
        
        if not alert:
            return None
        
        alert.acknowledged = True
        self.db.commit()
        self.db.refresh(alert)
        return alert

    # --- Auto Checks ---
    def auto_check_all(self) -> None:
        """Vérification automatique de tous les emails actifs (pour job background)"""
        emails = self.db.query(MonitoredEmail).filter(MonitoredEmail.active == True).all()
        
        for email in emails:
            # Check if it's time for a check based on frequency
            latest_check = self.get_latest_check(email.id, email.user_id)
            
            if latest_check:
                time_since_check = datetime.utcnow() - latest_check.checked_at
                
                if email.check_frequency == "daily" and time_since_check < timedelta(days=1):
                    continue
                elif email.check_frequency == "weekly" and time_since_check < timedelta(weeks=1):
                    continue
                elif email.check_frequency == "monthly" and time_since_check < timedelta(days=30):
                    continue
            
            try:
                self.check_email_security(email.id, email.user_id)
            except Exception as e:
                logger.error(f"Error auto-checking email {email.email}: {e}")

    # --- Stats ---
    def get_stats(self, user_id: int) -> Dict[str, Any]:
        """Statistiques de sécurité email"""
        total_monitored = self.db.query(MonitoredEmail).filter(MonitoredEmail.user_id == user_id).count()
        active_monitored = self.db.query(MonitoredEmail).filter(
            MonitoredEmail.user_id == user_id,
            MonitoredEmail.active == True
        ).count()
        
        # Count compromised emails
        compromised_emails = self.db.query(MonitoredEmail).join(EmailSecurityCheck).filter(
            MonitoredEmail.user_id == user_id,
            EmailSecurityCheck.is_compromised == True
        ).distinct().count()
        
        total_checks = self.db.query(EmailSecurityCheck).join(MonitoredEmail).filter(
            MonitoredEmail.user_id == user_id
        ).count()
        
        unacknowledged_alerts = self.db.query(EmailSecurityAlert).join(EmailSecurityCheck).join(MonitoredEmail).filter(
            MonitoredEmail.user_id == user_id,
            EmailSecurityAlert.acknowledged == False
        ).count()
        
        return {
            "total_monitored": total_monitored,
            "active_monitored": active_monitored,
            "compromised_emails": compromised_emails,
            "total_checks": total_checks,
            "unacknowledged_alerts": unacknowledged_alerts
        }

File: domains/informatique/02-projects/portfolio/Portfolio/angular-portfolio/src/app/core/services/email-security.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface MonitoredEmail {
  id: number;
  email: string;
  type: string;
  contact_id?: number;
  check_frequency: string;
  active: boolean;
  user_id: number;
  created_at: string;
}

export interface EmailSecurityCheck {
  id: number;
  monitored_email_id: number;
  checked_at: string;
  is_compromised: boolean;
  breach_count: number;
  breach_details?: Array<{
    hash_suffix: string;
    count: number;
  }>;
}

export interface EmailSecurityAlert {
  id: number;
  check_id: number;
  sent_at: string;
  acknowledged: boolean;
}

export interface EmailSecurityStats {
  total_monitored: number;
  active_monitored: number;
  compromised_emails: number;
  total_checks: number;
  unacknowledged_alerts: number;
}

export interface CreateMonitoredEmailRequest {
  email: string;
  type?: string;
  contact_id?: number;
  check_frequency?: string;
}

export interface UpdateMonitoredEmailRequest {
  type?: string;
  contact_id?: number;
  check_frequency?: string;
  active?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class EmailSecurityService {
  private apiUrl = `${environment.apiUrl}/email-security`;

  constructor(private http: HttpClient) {}

  // --- Monitored Emails ---
  createMonitoredEmail(email: CreateMonitoredEmailRequest): Observable<MonitoredEmail> {
    return this.http.post<MonitoredEmail>(`${this.apiUrl}/emails`, email);
  }

  getMonitoredEmails(activeOnly: boolean = false): Observable<MonitoredEmail[]> {
    let params = new HttpParams();
    if (activeOnly) {
      params = params.set('active_only', 'true');
    }
    return this.http.get<MonitoredEmail[]>(`${this.apiUrl}/emails`, { params });
  }

  getMonitoredEmail(emailId: number): Observable<MonitoredEmail> {
    return this.http.get<MonitoredEmail>(`${this.apiUrl}/emails/${emailId}`);
  }

  updateMonitoredEmail(emailId: number, email: UpdateMonitoredEmailRequest): Observable<MonitoredEmail> {
    return this.http.patch<MonitoredEmail>(`${this.apiUrl}/emails/${emailId}`, email);
  }

  deleteMonitoredEmail(emailId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/emails/${emailId}`);
  }

  checkEmailSecurity(emailId: number): Observable<EmailSecurityCheck> {
    return this.http.post<EmailSecurityCheck>(`${this.apiUrl}/emails/${emailId}/check`, {});
  }

  // --- Security Checks ---
  getSecurityChecks(emailId?: number, limit: number = 50): Observable<EmailSecurityCheck[]> {
    let params = new HttpParams();
    if (emailId) {
      params = params.set('email_id', emailId.toString());
    }
    params = params.set('limit', limit.toString());
    return this.http.get<EmailSecurityCheck[]>(`${this.apiUrl}/checks`, { params });
  }

  getLatestCheck(emailId: number): Observable<EmailSecurityCheck> {
    return this.http.get<EmailSecurityCheck>(`${this.apiUrl}/checks/latest/${emailId}`);
  }

  // --- Alerts ---
  getAlerts(acknowledged?: boolean): Observable<EmailSecurityAlert[]> {
    let params = new HttpParams();
    if (acknowledged !== undefined) {
      params = params.set('acknowledged', acknowledged.toString());
    }
    return this.http.get<EmailSecurityAlert[]>(`${this.apiUrl}/alerts`, { params });
  }

  acknowledgeAlert(alertId: number): Observable<EmailSecurityAlert> {
    return this.http.post<EmailSecurityAlert>(`${this.apiUrl}/alerts/${alertId}/acknowledge`, {});
  }

  // --- Stats ---
  getStats(): Observable<EmailSecurityStats> {
    return this.http.get<EmailSecurityStats>(`${this.apiUrl}/stats`);
  }
}
